dbms/src/Flash/Coprocessor/DAGExpressionAnalyzer.cpp

```diff
@@ -939,11 +939,16 @@ std::pair<bool, Names> DAGExpressionAnalyzer::buildJoinKey(
 
         String key_name = getActions(key, actions);
         DataTypePtr current_type = actions->getSampleBlock().getByName(key_name).type;
-        if (!removeNullable(current_type)->equals(*removeNullable(key_types[i])))
+        const auto & join_key_type = join_key_types[i];
+        if (!removeNullable(current_type)->equals(*removeNullable(join_key_type.key_type)))
         {
             /// need to convert to key type
-            key_name = appendCast(key_types[i], actions, key_name);
+            key_name = appendCast(join_key_type.key_type, actions, key_name);
             has_actions = true;
+            if (join_key_type.is_incompatible_decimal)
+            {
+                key_name = applyFunction("formatDecimalStr", {key_name}, actions, nullptr);
```

Comment:
Since we always have to add a new function to do the format, why not just add a function to format from deciml directly instead of convert decimal to string and format the result string?

Reply:
you're right, let me try :)

Reply:
done